cpp/src/decisiontree/batched-levelalgo/kernels.cuh

```diff
@@ -77,16 +77,24 @@ struct ClsDeviceTraits {
       atomicAdd(shist + label, 1);
     }
     __syncthreads();
-    auto op = Int2Max();
-    int2 v = {-1, -1};
-    for (int i = tid; i < input.nclasses; i += blockDim.x) {
-      int2 tmp = {i, shist[i]};
-      v = op(v, tmp);
+    int max_class_idx = 0;
+    int max_count = 0;
+    int total_count = 0;
+    for (int i = 0; i < input.nclasses; ++i) {
```

Comment:
This loop was executed collaboratively by threads in the block. Now it is executed redundantly by all the threads in the block. Any particular reasons for that?

Reply:
Two reasons:
1. I was following #3854, which also redundantly computes the sum over the classes.
2. This PR requires the computation of `total_count`, which is the sum of all elements of `shist`. If the loop were to run collaboratively, I'd need to define an extra data structure to obtain the correct value for `total_count`.

Reply:
`total_count` can be calculated with a call to `cub::BlockReduce`. Assuming `nclasses` is small this is small penalty. We could change it in future if that assumption breaks. 
In that case, can the whole loop be moved inside the `if (tid == 0)` block on line 91?

Reply:
@vinaydes Thanks. What do you think of #3854, where the summing is performed redundantly in all threads? For example: https://github.com/rapidsai/cuml/blob/bf5007c480917682269dc5a589b6b3aa8e0de126/cpp/src/decisiontree/batched-levelalgo/metrics.cuh#L112-L123

Reply:
If I understand correctly, in #3854, `LeafPrediction()` is called on line [kernels.cuh#L141](https://github.com/rapidsai/cuml/blob/bf5007c480917682269dc5a589b6b3aa8e0de126/cpp/src/decisiontree/batched-levelalgo/kernels.cuh#L141). It is already inside `tid == 0` block. I could be wrong if it is called at some other place too.

Reply:
@vinaydes Got it. In that case, I can move this loop inside the `tid == 0` block?